libs/candles-chart/src/lib/data-source.ts

```diff
@@ -213,6 +220,95 @@ export class VegaDataSource implements DataSource {
   }
 }
 
+const getDuration = (
+  interval: PennantInterval,
+  multiplier: number
+): Duration => {
+  switch (interval) {
+    case 'I1D':
+      return {
+        days: 1 * multiplier,
+      };
+    case 'I1H':
+      return {
+        hours: 1 * multiplier,
+      };
+    case 'I1M':
+      return {
+        minutes: 1 * multiplier,
+      };
+    case 'I5M':
+      return {
+        minutes: 5 * multiplier,
+      };
+    case 'I6H':
+      return {
+        hours: 6 * multiplier,
+      };
+    case 'I15M':
+      return {
+        minutes: 15 * multiplier,
+      };
+  }
+};
+
+const getDifference = (
+  interval: PennantInterval,
+  dateLeft: Date,
+  dateRight: Date
+): number => {
+  switch (interval) {
+    case 'I1D':
+      return differenceInDays(dateRight, dateLeft);
+    case 'I6H':
+      return differenceInHours(dateRight, dateLeft) / 6;
+    case 'I1H':
+      return differenceInHours(dateRight, dateLeft);
+    case 'I15M':
+      return differenceInMinutes(dateRight, dateLeft) / 15;
+    case 'I5M':
+      return differenceInMinutes(dateRight, dateLeft) / 5;
+    case 'I1M':
+      return differenceInMinutes(dateRight, dateLeft);
+  }
+};
+
+const checkGranulationContinuity =
+  (interval: PennantInterval) =>
+  (
+    agg: CandleFieldsFragment[],
+    candle: CandleFieldsFragment,
+    i: number
+  ): CandleFieldsFragment[] => {
+    if (agg.length && i) {
+      const previous = agg[agg.length - 1];
+      const previousStartDate = new Date(previous.periodStart);
+      const candleStartDate = new Date(candle.periodStart);
+      const difference = getDifference(
+        interval,
+        previousStartDate,
+        candleStartDate
+      );
+      if (difference > 1) {
+        for (let j = 1; j < difference; j++) {
+          const duration = getDuration(interval, j);
+          const newStartDate = add(previousStartDate, duration);
+          agg.push({
+            periodStart: newStartDate.toISOString(),
+            lastUpdateInPeriod: add(newStartDate, duration).toISOString(),
```

Comment:
☝🏻  that's not needed

Reply:
has been optimised in some different, much better way